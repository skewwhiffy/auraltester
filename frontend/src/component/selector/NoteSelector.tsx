import {useState} from "react"
import {Accidental, Note, NoteName} from "../../common/types"
import Centre from "../Centre"
import FancyRadioButtons from "../FancyRadioButtons"
import H1 from "../H1"
import NoteNameSelector from "./NoteNameSelector";

interface NoteSelectorProps {
    title?: string | null,
    value?: Note | null,
    showDoubles?: boolean

    onChange(note: Note): void
}

const NoteSelector = ({title, value, onChange, showDoubles = true}: NoteSelectorProps) => {
    const [noteName, setNoteName] = useState<NoteName | undefined>(value?.name)
    const [accidental, setAccidental] = useState<Accidental | undefined>(value?.accidental)

    const onNoteNameChange = (noteName: NoteName) => {
        setNoteName(noteName)
        if (accidental !== undefined && accidental !== null) {
            onChange({name: noteName, accidental})
        }
    }

    const onAccidentalChange = (accidental: Accidental) => {
        setAccidental(accidental)
        if (noteName !== undefined && noteName !== null) {
            onChange({name: noteName, accidental})
        }
    }

    return (
        <>
            {title && <Centre><H1>{title}</H1></Centre>}
            <NoteNameSelector value={noteName} onChange={onNoteNameChange}/>
            <Centre>
                <FancyRadioButtons
                    value={accidental}
                    onChange={onAccidentalChange}
                >
                    {showDoubles && <FancyRadioButtons.Item value='x'>Double sharp</FancyRadioButtons.Item>}
                    <FancyRadioButtons.Item>#</FancyRadioButtons.Item>
                    <FancyRadioButtons.Item value=''>Natural</FancyRadioButtons.Item>
                    <FancyRadioButtons.Item>b</FancyRadioButtons.Item>
                    {showDoubles && <FancyRadioButtons.Item value='bb'>Double flat</FancyRadioButtons.Item>}
                </FancyRadioButtons>
            </Centre>
        </>
    )
}

export default NoteSelector