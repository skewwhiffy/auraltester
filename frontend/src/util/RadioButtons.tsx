import React from 'react'
import { Form } from "react-bootstrap";

type OnChangeHandler = (value: string) => void

export interface RadioButtonDefinition {
  key?: string,
  value: string,
  label?: string
}

export interface Props {
  values: Array<RadioButtonDefinition>,
  name: string,
  defaultValue?: string,
  onChange: OnChangeHandler
}

interface State {
  value: string
}

class RadioButtons extends React.Component<Props, State> {
  constructor(props: Props) {
    if (props.values.length === 0) {
      throw Error('Need at least one radio button')
    }
    super(props);
  }

  private onChange = (ev: any) => {
    (this.props.onChange || (_ => { }))(ev.target.value)
  }

  private renderRadioButton = (radioButton: RadioButtonDefinition) => {
    const value = radioButton.value
    const key = radioButton.key || value
    const label = radioButton.label || value
    const defaultValue = this.props.defaultValue === undefined
      ? this.props.values[0].value
      : this.props.defaultValue
    return (
      <Form.Check
        id={`${this.props.name}-${key}`}
        key={key}
        name={this.props.name}
        value={value}
        label={label}
        type='radio'
        defaultChecked={value === defaultValue}
      />
    )
  }

  render() {
    return (
      <Form.Group onChange={this.onChange}>
        {this.props.values.map(this.renderRadioButton)}
      </Form.Group>
    )
  }
}

export default RadioButtons