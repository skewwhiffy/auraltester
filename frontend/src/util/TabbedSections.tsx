import { Container, Nav } from "react-bootstrap"
import { Route, Routes } from "react-router"
import { Link } from "react-router-dom"
import { capitalizeFirstCharacter } from "."

export type TabbedSection = {
  getElement: () => JSX.Element
} & ({ path: string } | { text: string })

export interface Props {
  sections: TabbedSection[]
}

const getNavLink = (it: TabbedSection): JSX.Element => {
  const to = "path" in it ? it.path : "."
  const text = "text" in it ? it.text : capitalizeFirstCharacter(it.path)
  return (
    <Nav.Item key={to}>
      <Nav.Link as={Link} to={to}>{text}</Nav.Link>
    </Nav.Item>
  )
}

const getRoute = (it: TabbedSection): JSX.Element => {
  return "path" in it
    ? <Route key={it.path} path={it.path} element={it.getElement()} />
    : <Route key="." index element={it.getElement()} />
}

const TabbedSections = (props: Props): JSX.Element => {
  return (
    <Container>
      <Nav variant="tabs">
        {props.sections.map(getNavLink)}
      </Nav>
      <Routes>
        {props.sections.map(getRoute)}
      </Routes>
    </Container >
  )
}

export default TabbedSections