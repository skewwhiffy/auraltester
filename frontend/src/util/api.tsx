import axios from "axios"

interface ClefResponse {
  abc: string
}

const api = {
  async getClef(clef: string): Promise<ClefResponse> {
    const response = await axios.get("/api/clef", {
      params: {
        clef
      }
    })
    return response.data
  }
}

export default api