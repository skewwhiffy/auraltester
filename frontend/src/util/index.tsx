import api from './api'

const capitalizeFirstCharacter = (source: string): string => source.length === 0
  ? ''
  : `${source[0].toUpperCase()}${source.substring(1)}`

export { capitalizeFirstCharacter, api }
