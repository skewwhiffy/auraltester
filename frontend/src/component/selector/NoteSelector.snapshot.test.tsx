import NoteSelector from './NoteSelector';
import { Props as RadioButtonProps } from '../../util/RadioButtons'
import renderer from 'react-test-renderer'

jest.mock('../../util/RadioButtons', () => (props: RadioButtonProps) => (
  <div>{JSON.stringify(props)}</div>
))

describe('NoteSelector snapshot', () => {
  it('matches snapshot', () => {
    const actual = renderer.create(
      <NoteSelector defaultValue='C' includeDoubleAccidentals onChange={() => { }} />
    )

    expect(actual).toMatchSnapshot()
  })

  it('matches snapshot with no includeDoubleAccidentals', () => {
    const actual = renderer.create(
      <NoteSelector defaultValue='D' onChange={() => { }} />
    )

    expect(actual).toMatchSnapshot()
  })
})