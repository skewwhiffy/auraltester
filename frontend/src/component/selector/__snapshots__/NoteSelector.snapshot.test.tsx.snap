// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NoteSelector snapshot matches snapshot 1`] = `
<div
  className="container"
>
  <div
    className="row"
  >
    <div
      className="col"
    >
      Note
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="col"
    >
      <div>
        {"values":[{"value":"A"},{"value":"B"},{"value":"C"},{"value":"D"},{"value":"E"},{"value":"F"},{"value":"G"}],"name":"note","defaultValue":"C"}
      </div>
    </div>
    <div
      className="col"
    >
      <div>
        {"values":[{"value":"x","label":"double sharp"},{"value":"#"},{"value":"","label":"natural"},{"value":"b"},{"value":"bb","label":"double flat"}],"name":"accidental","defaultValue":""}
      </div>
    </div>
  </div>
</div>
`;

exports[`NoteSelector snapshot matches snapshot with no includeDoubleAccidentals 1`] = `
<div
  className="container"
>
  <div
    className="row"
  >
    <div
      className="col"
    >
      Note
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="col"
    >
      <div>
        {"values":[{"value":"A"},{"value":"B"},{"value":"C"},{"value":"D"},{"value":"E"},{"value":"F"},{"value":"G"}],"name":"note","defaultValue":"D"}
      </div>
    </div>
    <div
      className="col"
    >
      <div>
        {"values":[{"value":"#"},{"value":"","label":"natural"},{"value":"b"}],"name":"accidental","defaultValue":""}
      </div>
    </div>
  </div>
</div>
`;
